dali/python/nvidia/dali/external_source.py

```diff
@@ -0,0 +1,244 @@
+"""
+    _args_doc = """
+Args
+----
+
+`source` : callable or iterable
+    The source of the data. The source is polled for data (via a call `source()` or `next(source)`
+    whenever the pipeline needs input for the next iteration. The source can supply one or more data
+    batches, depending on the value of `num_outputs`. If `num_outputs` is not set, the `source` is
+    expected to return a single batch. If it's specified, the data is expected to a be tuple or list
+    where each element corresponds to respective return value of the external_source.
+    If the source is a callable and has a positional argument, it is assumed to be the current
+    iteration number and consecutive calls will be `source(0)`, `source(1)`, etc.
+
+`num_outputs` : int, optional
+    If specified, denotes the number of TensorLists produced by the source function
+
+Keyword Args
+------------
+
+`cycle`: bool
+    If `True`, the source iterable will be wrapped. Otherwise, StopIteration error wil be raised
+    when end of data is reached. Setting this flag to True when `source` is not an iterable is an
+    error.
+
+`name` : str, optional
+    The name of the data node - used when feeding the data in `iter_setup`; can be omitted if
+    the data is provided by `source`.
+
+`layout` : str or list/tuple of str:
+    If provided, sets the layout of the data. When `num_outputs` > 1, layout can be a list
+    containing a distinct layout for each output. If the list has fewer elements than `num_outputs`,
+    only the first outputs have the layout set, the reset have it cleared.
+"""
+
+    def __init__(self, source = None, num_outputs = None, *, cycle = False, layout = None, name = None, device = "cpu", **kwargs):
+        self._schema = _b.GetSchema("_ExternalSource")
+        self._spec = _b.OpSpec("_ExternalSource")
+        self._device = device
+        self._layout = layout
+
+        callback = _get_callback_from_source(source, cycle)
+
+        if name is not None and self.num_outputs is not None:
+            raise ValueError("`num_outputs` is not compatible with named `ExternalSource`")
+
+        self._name = name
+        self._num_outputs = num_outputs
+        self._callback = callback
+
+        self._spec.AddArg("device", device)
+        for key, value in kwargs.items():
+            self._spec.AddArg(key, value)
+
+    @property
+    def spec(self):
+        return self._spec
+
+    @property
+    def schema(self):
+        return self._schema
+
+    @property
+    def device(self):
+        return self._device
+
+    @property
+    def preserve(self):
+        return False
+
+    def __call__(self, *, source = None, cycle = None, name = None, layout = None, **kwargs):
+        ""
+        from nvidia.dali.ops import _OperatorInstance
+
+        if source is None:
+            if cycle is not None:
+                raise ValueError("The argument `cycle` can only be specified if `source` is iterable")
```

Comment:
It must actually be iterable - cycle doesn't make sense for callbacks.

Reply:
But in this context you are just checking if the user didn't pass the source to call operator.

Reply:
This would be a very bad user experience of iteratively fixing stuff:
no source, cycle = True -> "source not specified"
source = callable, cycle = True -> "another error??"
why not do it in one go?

Reply:
But what if:
_callback = lambda, source = None, cycle = True
The error would indicate that something is wrong with the callable that was passed to init, and not about passing some arguments to init and other to call.